Add metrics section to benchmark (#4972)
opensearch-project/documentation-website: _benchmark/metrics/metric-records.md

---
layout: default
title: Metric records
nav_order: 30
parent: Metrics
---

# Metric records

> OpenSearch Benchmark stores metrics in the `benchmark-metrics-*` indices. A new index is created each month. The following is an example metric record stored in the `benchmark-metrics-2023-08` index:

Reviewer: ```suggestion
OpenSearch Benchmark stores metrics in the `benchmark-metrics-*` indexes. A new index is created each month. The following is an example metric record stored in the `benchmark-metrics-2023-08` index:
```

    "workload": "geonames",
    "test_procedure": "append-no-conflicts",
    "provision-config-instance": "external",
    "name": "service_time",
    "value": 607.8001195564866,
    "unit": "ms",
    "sample-type": "normal",
    "meta": {
      "source_revision": "unknown",
      "distribution_version": "1.1.0",
      "distribution_flavor": "oss",
      "index": "geonames",
      "took": 13,
      "success": true,
      "success-count": 125,
      "error-count": 0
    },
    "task": "index-append",
    "operation": "index-append",
    "operation-type": "bulk"
  },
  "fields": {
    "@timestamp": [
      "2023-08-10T21:27:22.821Z"
    ],
    "test-execution-timestamp": [
      "2023-08-10T21:27:11.000Z"
    ]
  },
  "highlight": {
    "workload": [
      "@opensearch-dashboards-highlighted-field@geonames@/opensearch-dashboards-highlighted-field@"
    ],
    "meta.index": [
      "@opensearch-dashboards-highlighted-field@geonames@/opensearch-dashboards-highlighted-field@"
    ]
  },
  "sort": [
    1691702831000
  ]
}
```

The following fields found in the `_source` section of the metric's record are configurable in the `opensearch-benchmarks-metrics-*` file.

## @timestamp

> The timestamp when the sample was taken since the epoch, in milliseconds. For request-related metrics such as `latency` or `service_time`, this is the timestamp when OpenSearch Benchmark has issued the request.

Reviewer: ```suggestion
The timestamp of when the sample was taken since the epoch, in milliseconds. For request-related metrics, such as `latency` or `service_time`, this is the timestamp of when OpenSearch Benchmark issued the request.
```

## relative-time-ms

> The relative time since the start of the benchmark, in milliseconds. This is useful for comparing time series graphs over multiple tests. For example, you can compare the indexing throughput over time across multiple tests. 

Reviewer: ```suggestion
The relative time since the start of the benchmark, in milliseconds. This is useful for comparing time-series graphs across multiple tests. For example, you can compare the indexing throughput over time across multiple tests. 
```

## test-execution-id

> A UUID which changes on every invocation of the workload. It is intended to group all samples of a benchmarking run.

Reviewer: ```suggestion
A UUID that changes on every invocation of the workload. It is intended to group all samples of a benchmarking run.
```

## test-execution-timestamp

> The timestamp when the workload was invoked (always in UTC).

Reviewer: ```suggestion
The timestamp of when the workload was invoked (always in UTC).
```

## environment

> The `environment` describes the origin of a metric record. This is defined when initially [configuring]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/) OpenSearch Benchmark. You can use separate environments for different Benchmarks but store the metric records in the same index.

Reviewer: ```suggestion
The `environment` describes the origin of a metric record. This is defined when initially [configuring]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/) OpenSearch Benchmark. You can use separate environments for different benchmarks but store the metric records in the same index.
```

> The environment describes the origin of a metric record. This is defined when initially [configuring]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark) OpenSearch Benchmark. You can use seperate environments for different Benchmarks, but store the metric records in the same index.

Author: ```suggestion
The environment describes the origin of a metric record. This is defined when initially [configuring]({{site.url}}{{site.baseurl}}/benchmark/configuring-benchmark/) OpenSearch Benchmark. You can use separate environments for different Benchmarks, but store the metric records in the same index.
```

## workload, test_procedure, provision-config-instance

The workload, test procedures, and the configuration instances for which the metrics are produced.

## name, value, unit

> The actual metric name and value with an optional unit. Depending on the nature of a metric, it is either sampled periodically by OpenSearch Benchmark, for example, CPU utilization or query latency, or measured once, for example, the final size of the index.

Reviewer: ```suggestion
The actual metric name and value, with an optional unit. Depending on the nature of a metric, it is either sampled periodically by OpenSearch Benchmark, for example, CPU utilization or query latency, or measured once, for example, the final size of the index.
```

## sample-type

> Determines whether to configure a benchmark to run in warmup mode by setting to `warmup` or `normal`. Only `normal` samples are considered for the results that are reported.

Reviewer: ```suggestion
Determines whether to configure a benchmark to run in warmup mode by setting it to `warmup` or `normal`. Only `normal` samples are considered for the results that are reported.
```

## meta

The meta information for each metric record, including

> - CPU info: The number of physical and logical cores and also the model name

Reviewer 2: ```suggestion
- CPU info: The number of physical and logical cores and the model name.
```

> - OS info: OS name and version

Author: ```suggestion
- OS info: The name and version of the operating system.
```

> - Host name.

Author: ```suggestion
- Hostname.
```

> - Host name

Reviewer 2: ```suggestion
- Host name.
```

> - Node name: A unique name for each node when OpenSearch Benchmark provisions the cluster.

Reviewer: ```suggestion
- Node name: A unique name given to each node when OpenSearch Benchmark provisions the cluster.
```

> - Source revision: The git hash of the version of OpenSearch that is benchmarked. 

Reviewer: Should "Git" be capitalized?

> - Distribution version: The distribution version of Elasticsearch that is benchmarked. 

Reviewer 2: Elasticsearch?

> - Distribution version: The distribution version of Elasticsearch that is benchmarked. 

Author: ```suggestion
- Distribution version: The distribution version of OpenSearch that is benchmarked. 
```

> - Custom tags: You can define custom tags with the command line flag `--user-tags`. The tags are prefixed by `tag_` in order to avoid accidental clashes with OpenSearch Benchmark internal tags.

Reviewer: Does the first instance of "with" mean "by using"?

> - Custom tags: You can define custom tags with the command line flag `--user-tags`. The tags are prefixed by `tag_` in order to avoid accidental clashes with OpenSearch Benchmark internal tags.

Author: ```suggestion
- Custom tags: You can define custom tags by using the command line flag `--user-tags`. The tags are prefixed by `tag_` in order to avoid accidental clashes with OpenSearch Benchmark internal tags.
```

> - Custom tags: You can define custom tags with the command line flag --user-tags. The tags are prefixed by tag_ in order to avoid accidental clashes with OpenSearch Benchmark's internal tags.

Author: ```suggestion
- Custom tags: You can define custom tags with the command line flag `--user-tags`. The tags are prefixed by `tag_` in order to avoid accidental clashes with OpenSearch Benchmark's internal tags.
```

> - Custom tags: You can define custom tags with the command line flag `--user-tags`. The tags are prefixed by `tag_` in order to avoid accidental clashes with OpenSearch Benchmark's internal tags.

Reviewer 2: ```suggestion
- Custom tags: You can define custom tags with the command line flag `--user-tags`. The tags are prefixed by `tag_` in order to avoid accidental clashes with OpenSearch Benchmark internal tags.
```

> - Operation-specific: An optional substructure of the operation. For bulk requests, this may be the number of documents; for searches, the number of hits.

Reviewer: ```suggestion
- Operation specific: An optional substructure of the operation. For bulk requests, this may be the number of documents; for searches, the number of hits.
```

> - Operation-specific: The optional substructure operation contains additional information depending on the type of operation. For bulk requests, this may be the number of documents or for searches the number of hits.

Reviewer 2: Is there a noun missing here? Operation-specific fields? Also, I see in the response the `operation` field but it's a string and not an object. So what is the substructure?
```suggestion
- Operation-specific: The optional `operation` structure contains additional information, depending on the type of operation. For bulk requests, this may be the number of documents; for searches, the number of hits.
```

> Depending on the level of metric record, some meta information might be missing.

Reviewer 2: Could you add some more details on what the level means?

> Depending on the level of metric record, some meta information might be missing.

Author: ```suggestion
Depending on the metric record, some meta information might be missing.
```

## Next steps

> - for more information about how to access OpenSearch Benchmark metrics, see [Metrics]({{site.url}}{{site.baseurl}}/benchmark/metrics/index/).

Author: ```suggestion
- For more information about how to access OpenSearch Benchmark metrics, see [Metrics]({{site.url}}{{site.baseurl}}/benchmark/metrics/index/).
```

> - For more information about the metrics stored in OpenSearch Benchmark, see [Metric keys]({{site.url}}{{site.baseurl}}/benchmark/metrics/metric-keys).

Author: ```suggestion
- For more information about the metrics stored in OpenSearch Benchmark, see [Metric keys]({{site.url}}{{site.baseurl}}/benchmark/metrics/metric-keys/).
```
